core/src/main/java/org/elasticsearch/index/query/InnerHitBuilder.java

```diff
@@ -101,7 +101,7 @@
             } catch (IOException e) {
                 throw new ParsingException(p.getTokenLocation(), "Could not parse inner _source definition", e);
             }
-        }, SearchSourceBuilder._SOURCE_FIELD, ObjectParser.ValueType.OBJECT_OR_BOOLEAN);
+        }, SearchSourceBuilder._SOURCE_FIELD, ObjectParser.ValueType.OBJECT_ARRAY_OR_STRING);
```

Comment:
Not a big deal but this would break the parsing as a boolean eg.`_source:true`. You can maybe add a OBJECT_ARRAY_BOOLEAN_OR_STRING to the enum ?